thoth-api/src/model/location/crud.rs

```diff
@@ -126,7 +126,98 @@ impl Crud for Location {
         crate::model::publication::Publication::from_id(db, &self.publication_id)?.publisher_id(db)
     }
 
-    crud_methods!(location::table, location::dsl::location);
+    // `crud_methods!` cannot be used for update(), because we need to execute multiple statements
+    // in the same transaction for changing a non-canonical location to canonical.
+    // These functions recreate the `crud_methods!` logic.
+    fn from_id(db: &crate::db::PgPool, entity_id: &Uuid) -> ThothResult<Self> {
+        let mut connection = db.get()?;
+        match location::table
+            .find(entity_id)
+            .get_result::<Self>(&mut connection)
+        {
+            Ok(t) => Ok(t),
+            Err(e) => Err(ThothError::from(e)),
+        }
+    }
+
+    fn create(db: &crate::db::PgPool, data: &NewLocation) -> ThothResult<Self> {
+        let mut connection = db.get()?;
+
+        connection.transaction(|connection| {
+            diesel::insert_into(location::table)
+                .values(data)
+                .get_result::<Self>(connection)
+                .map_err(|e| e.into())
+        })
+    }
+
+    fn update(
```

Comment:
Hey, this is all great, and thanks for the streamlining! It appears that we still need to clone `old_canonical_location.location_id`; when I remove `clone()`, I get a compiler error, "cannot move out of `old_canonical_location` because it is borrowed". So I put the `clone()` back in.

Reply:
Disregard - I got it to work without using `clone()`! Not sure what was going on.